flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java

```diff
@@ -210,12 +210,15 @@ public int getInputIndex() {
 			final InputChannel channel = checkpointedInputGate.getChannel(channelIndex);
 
 			// Assumption for retrieving buffers = one concurrent checkpoint
-			recordDeserializers[channelIndex].getUnconsumedBuffer().ifPresent(buffer ->
-				channelStateWriter.addInputData(
-					checkpointId,
-					channel.getChannelInfo(),
-					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
-					buffer));
+			RecordDeserializer<?> deserializer = recordDeserializers[channelIndex];
+			if (deserializer != null) {
```

Comment:
I guess the reason of not using `Optional.flatMap` was `IOException` thrown from `RecordDeserializer.getUnconsumedBuffer`.

The root cause of that `IOException` is `DataOutputSerializer` wrapping with it OOMs and RuntimeExceptions.

Since the resulting `IOException` is never recovered or analyzed and originally it's RuntimeException I'd consider wrap `IOException` and use `flatMap`.

Reply:
Yes, I'm not a big fan of using `IOException`s everywhere, especially for unrecoverable issues. So your suggestion makes lots of sense.
But I'd also like to avoid larger refactorings right now (and `DataOutputSerializer` is used at various places), when it can be easily avoided as above. I'm assuming the quest for a more functional style takes a long time with the current exception handling as is.